website/www/site/content/en/blog/apache-beam-flink-and-kubernetes-part2.md

```diff
@@ -0,0 +1,161 @@
+1. **Stopping New Data:** The job stops reading new input.
+2. **Marking the Source:** It marks the source with an 'infinite watermark'. This is a bit technical, but think of it as a marker that tells the system there's no more new data to process.
+3. **Propagation Through the Pipeline:** This marker is then passed along through the job's processing pipeline, ensuring every part of the job knows not to expect any new data.
+
+This seemingly small change has a big impact. It ensures that when a job is updated or deleted, the data it processes remains consistent and accurate. This is crucial for any data-processing task, as it maintains the integrity and reliability of the data. Furthermore, in cases where the drainage fails, the job can be canceled without needing a savepoint, which adds a layer of flexibility and safety to the whole process.
+
+# Conclusion
+
+As we conclude Part 2 of our series on building and managing Apache Beam Flink services on Kubernetes, it's evident that the journey of implementing autoscaling has been both challenging and enlightening. The obstacles we faced, from understanding Apache Beam backlog metrics in the Flink Runner environment to addressing slow reads in high-traffic scenarios, pushed us to develop innovative solutions and deepen our understanding of streaming infrastructure.
+
+Our exploration into the intricacies of checkpointing, Kafka reader wait times, and unbalanced partition distribution revealed the complexities of autoscaling Beam jobs. These challenges prompted us to devise strategies like the adaptive timeout in KafkaIO and the balanced workload distribution in Flink's SlotSharingSlotAllocator. Additionally, the introduction of the drain support in Kubernetes Operator with Flink marks a significant advancement in managing stateful and stateless jobs effectively.
+
+This journey has not only enhanced the robustness and efficiency of our system but has also contributed valuable insights to the broader community working with Apache Beam and Flink. We hope that our experiences and solutions will aid others facing similar challenges in their projects.
+
+Stay tuned for our next blog post where we will delve into the specifics of Autoscaling in Apache Beam. We'll break down the concepts, strategies, and best practices to effectively scale your Beam jobs. Thank you for following our series, and we look forward to sharing more of our journey and learnings with you.
+
+# Acknowledgements
+
+This is a large effort to build the new infrastructure and to migrate the large customer based applications from cloud provider managed streaming infrastructure to self-managed Flink based infrastructure at scale. Thanks the Palo Alto Networks CDL streaming team who helped to make this happen: Kishore Pola, Andrew Park, Hemant Kumar, Manan Mangal, Helen Jiang, Mandy Wang, Praveen Kumar Pasupuleti, JM Teo, Rishabh Kedia, Talat Uyarer, Naitk Dani, and David He.
```

Comment:
```suggestion
This is a large effort to build the new infrastructure and to migrate the large customer based applications from cloud provider managed streaming infrastructure to self-managed, Flink-based infrastructure at scale. Thanks the Palo Alto Networks CDL streaming team who helped to make this happen: Kishore Pola, Andrew Park, Hemant Kumar, Manan Mangal, Helen Jiang, Mandy Wang, Praveen Kumar Pasupuleti, JM Teo, Rishabh Kedia, Talat Uyarer, Naitk Dani, and David He.
```

Reply:
Resolving the outdated comment. Thank you for your help!